### PR TITLE
Fix for #2686 - cache misses / bad authority validation on PPE + regi…

### LINKED ISSUE #2686: [Bug] Cache misses in client_credentials when authority is specified at request level, cloud is not global and region is used
```csharp
 var cca = ConfidentialClientApplicationBuilder.Create("client_id")              
              .WithClientSecret("secret")
              .WithAzureRegion("eastus")            
              .Build();

            var result = await cca.AcquireTokenForClient(new[] { "https://graph.ppe.windows.net/.default" })
                           .WithAuthority("https://login.windows-ppe.net/17b189bc-2b81-4ec5-aa51-3e628cbc931b")
                           .ExecuteAsync();

            Assert.AreEqual(TokenSource.IdentityProvider, result.AuthenticationResultMetadata.TokenSource);

            var result2 = await cca.AcquireTokenForClient(new[] { "https://graph.ppe.windows.net/.default" })
                       .WithAuthority("https://login.windows-ppe.net/17b189bc-2b81-4ec5-aa51-3e628cbc931b")
                       .ExecuteAsync();

            // fails
            Assert.AreEqual(TokenSource.Cache, result2.AuthenticationResultMetadata.TokenSource);
```

**Workaround**: Add `.WithAuthority("https://login.windows-ppe.net/common")` or any authority of the same cloud **to the application builder.**

```csharp

 var cca = ConfidentialClientApplicationBuilder.Create("client_id")              
              .WithClientSecret("secret")
              .WithAzureRegion("eastus")            
              .WithAuthority("https://login.windows-ppe.net/common")              
              .Build();

```

### PR DESCRIPTION
**Pull request by bgavrilMS**
…onal

Fixes #2686  .

**Changes proposed in this request**
- prevent developers from 
- force validation of non-regionalized env for unknown environments like PPE

**Testing**
Unit tests
I have an int test with PPE as well, will publish it as a separate PR because I need to hide secrets in KV etc. Test runs fine on my machine with a secret in clear.

**Performance impact**
fingers crossed.... there should be no impact on known clouds

### CODE REVIEW

File: src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs [hunk 1]
```diff
@@ -532,6 +534,11 @@ private async Task<IEnumerable<MsalAccessTokenCacheItem>> FilterByEnvironmentAsy
 
             if (filteredByPreferredAlias.Any())
             {
+                if (logger.IsLoggingEnabled(LogLevel.Verbose))
+                {
+                    logger.Verbose($"Filtered by preffered alias returning {filteredByPreferredAlias.Count()} tokens");
```
Comment (jennyf19) [inline src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs:539]:
```suggestion
                    logger.Verbose($"Filtered by preferred alias returning {filteredByPreferredAlias.Count()} tokens");
```

File: src/client/Microsoft.Identity.Client/Instance/Authority.cs [hunk 2]
```diff
@@ -195,3 +214,13 @@ internal static string GetEnvironment(string authority)
         }
     }
 }
+
+/// <summary>
```
Comment (jennyf19) [inline src/client/Microsoft.Identity.Client/Instance/Authority.cs:218]: delete this or move it above the method?

File: tests/Microsoft.Identity.Test.Unit/AuthorityAliasesTests.cs [hunk 3]
```diff
@@ -48,11 +46,14 @@ public async Task AuthorityMigrationTestAsync()
                     .Create(TestConstants.ClientId)
                           .WithAuthority(authorityUri, true)
                           .WithHttpManager(httpManager)
-                          .WithUserTokenLegacyCachePersistenceForTest(new TestLegacyCachePersistance())
+                          //.WithUserTokenLegacyCachePersistenceForTest(new TestLegacyCachePersistance())
```
Comment (jennyf19) [inline tests/Microsoft.Identity.Test.Unit/AuthorityAliasesTests.cs:49]: remove if not needed or add a comment for when it should be used?

File: tests/Microsoft.Identity.Test.Unit/AuthorityAliasesTests.cs [hunk 4]
```diff
@@ -108,7 +120,18 @@ public async Task AuthorityMigrationTestAsync()
 
                     try
                     {
-                        result = await app
+                        var app3 = PublicClientApplicationBuilder
+                               .Create(TestConstants.ClientId)
+                                     .WithAuthority($"https://{envAlias}/common", true)
```
Comment (jennyf19) [inline tests/Microsoft.Identity.Test.Unit/AuthorityAliasesTests.cs:125]: >true [](http://example.com/codeflow?start=82&length=4)

what if .WithAuthority("", false)?

**Review by jennyf19 — APPROVED** [review]
:shipit:

File: tests/Microsoft.Identity.Test.Unit/AuthorityAliasesTests.cs [hunk 5]
```diff
@@ -76,7 +77,18 @@ public async Task AuthorityMigrationTestAsync()
                 // silent request targeting at, should return at from cache for any environment alias
                 foreach (var envAlias in TestConstants.s_prodEnvAliases)
                 {
-                    result = await app
+                    var app2 = PublicClientApplicationBuilder
+                    .Create(TestConstants.ClientId)
+                          .WithAuthority($"https://{envAlias}/common", true)
+                          .WithHttpManager(httpManager)
+                          //.WithUserTokenLegacyCachePersistenceForTest(new TestLegacyCachePersistance())
```
Comment (trwalke) [inline tests/Microsoft.Identity.Test.Unit/AuthorityAliasesTests.cs:84]: dead code?

File: tests/Microsoft.Identity.Test.Unit/ApiConfigTests/AuthorityTests.cs [hunk 6]
```diff
@@ -13,6 +13,9 @@ public class AuthorityTests
     {
         private static readonly AuthorityInfo s_commonAuthority =
             AuthorityInfo.FromAuthorityUri(TestConstants.AuthorityCommonTenant, true);
+        static string s = $@"https://{TestConstants.PPEEnvironment}/{TestConstants.TenantId}";
```
Comment (pmaytak) [inline tests/Microsoft.Identity.Test.Unit/ApiConfigTests/AuthorityTests.cs:16]: nit: maybe move this to TestConstants or rename.

File: tests/Microsoft.Identity.Test.Unit/AuthorityAliasesTests.cs (hunk 4 above)
Comment (bgavrilMS) [reply tests/Microsoft.Identity.Test.Unit/AuthorityAliasesTests.cs:124]: Added all 4 combinations of tests now: 

validateAuthority = true | false
isAuthorityValid = true | false

File: src/client/Microsoft.Identity.Client/Instance/Discovery/InstanceDiscoveryMetadataEntry.cs [hunk 7]
```diff
@@ -9,6 +9,10 @@ namespace Microsoft.Identity.Client.Instance.Discovery
     [Preserve(AllMembers = true)]
     internal sealed class InstanceDiscoveryMetadataEntry
     {
+        public InstanceDiscoveryMetadataEntry()
```
[the remark this answers is not in the document]
Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/Instance/Discovery/InstanceDiscoveryMetadataEntry.cs:12]: I wanted to see where we create this object. The compiler generates a parameterless ctor in any case.

File: src/client/Microsoft.Identity.Client/Instance/Discovery/InstanceDiscoveryManager.cs [hunk 8]
```diff
@@ -121,9 +122,17 @@ internal class InstanceDiscoveryManager : IInstanceDiscoveryManager
             {
                 case AuthorityType.Aad:
 
-                    var entry = _userMetadataProvider?.GetMetadataOrThrow(environment, requestContext.Logger) ??
-                                await _regionDiscoveryProvider.GetMetadataAsync(authorityUri, requestContext).ConfigureAwait(false) ??
-                                await FetchNetworkMetadataOrFallbackAsync(requestContext, authorityUri).ConfigureAwait(false);
+                    var entry = _userMetadataProvider?.GetMetadataOrThrow(environment, requestContext.Logger);
+                    InstanceDiscoveryMetadataEntry networkEntry = null;
+                    if (entry == null && forceValidation)
```
[the remark this answers is not in the document]
Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/Instance/Discovery/InstanceDiscoveryManager.cs:127]: I think I need to redesign this part, it's hard to read / maintain. In this particular case, the _userMetadataProvider would be null.

**Pull request merged**